### PR TITLE
Complete project example

### PR DESCRIPTION
**Pull request by pdan93**
Complete project that exemplifies implementing custom classifiers (both svm and transformers)

### CODE REVIEW

**Review by ivyleavedtoadflax — COMMENTED** [review]
Are these files meant to be here?

custom_model/runs/Jan19_14-18-03_daniel-RM-DESKTOP-ECOQUIET-320/1642594687.8485427/events.out.tfevents.1642594687.daniel-RM-DESKTOP-ECOQUIET-320.14899.1
custom_model/runs/Jan19_14-18-03_daniel-RM-DESKTOP-ECOQUIET-320/events.out.tfevents.1642594687.daniel-RM-DESKTOP-ECOQUIET-320.14899.0

Also I thought this was about sklearn, not about transformers? Is the plan to publish a transformers example too?

File: config.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# https://rasa.com/docs/rasa/nlu/components/
+language: en
+
+pipeline:
+   #- name: sklearn_classifier.CustomIntentSVM
+   #  kernel: "linear"
+   #  coef0: 1
+   #  degree: 2
+   #  class_weight: "balanced"
+   - name: transformer_classifier.TransfromerClassifier
```
Comment (nsorros) [inline config.yml]: typo: `TransfromerClassifier`

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+rasa==2.8.0
```
Comment (nsorros) [inline requirements.txt:1]: I think I prefer pinned requirements vs the unpinned to pinned situation

**Comment by nsorros** [comment]
`rasa train` on a new env breaks
```
2022-01-20 09:11:14 INFO     rasa.nlu.model  - Finished training component.
tokenizer config file saved in /tmp/tmpwlbb5tbm/nlu/tokenizer_component_0_transformer_classifier/tokenizer_config.json
Special tokens file saved in /tmp/tmpwlbb5tbm/nlu/tokenizer_component_0_transformer_classifier/special_tokens_map.json
Configuration saved in /tmp/tmpwlbb5tbm/nlu/model_component_0_transformer_classifier/config.json
Model weights saved in /tmp/tmpwlbb5tbm/nlu/model_component_0_transformer_classifier/pytorch_model.bin
Configuration saved in /tmp/tmpwlbb5tbm/nlu/config_component_0_transformer_classifier/config.json
2022-01-20 09:11:14 INFO     rasa.nlu.model  - Successfully saved model into '/tmp/tmpwlbb5tbm/nlu'
NLU model training completed.
Traceback (most recent call last):
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/bin/rasa", line 8, in <module>
    sys.exit(main())
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/__main__.py", line 117, in main
    cmdline_arguments.func(cmdline_arguments)
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/cli/train.py", line 59, in <lambda>
    train_parser.set_defaults(func=lambda args: run_training(args, can_exit=True))
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/cli/train.py", line 91, in run_training
    training_result = train_all(
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/api.py", line 109, in train
    return rasa.utils.common.run_in_loop(
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/utils/common.py", line 296, in run_in_loop
    result = loop.run_until_complete(f)
  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 108, in train_async
    return await _train_async_internal(
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 288, in _train_async_internal
    await _do_training(
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 358, in _do_training
    interpreter=_load_interpreter(interpreter_path)
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 382, in _load_interpreter
    return rasa.core.interpreter.create_interpreter(interpreter_path)
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/core/interpreter.py", line 33, in create_interpreter
    return RasaNLUInterpreter(model_directory=obj)
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/core/interpreter.py", line 127, in __init__
    self._load_interpreter()
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/core/interpreter.py", line 165, in _load_interpreter
    self.interpreter = Interpreter.load(self.model_directory)
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/model.py", line 341, in load
    return Interpreter.create(
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/model.py", line 416, in create
    component = component_builder.load_component(
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/components.py", line 902, in load_component
    component = registry.load_component_by_meta(
  File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/registry.py", line 177, in load_component_by_meta
    return component_class.load(
  File "/root/code/rasa_custom_intent_classification/transformer_classifier.py", line 255, in load
    x = cls()
  File "/root/code/rasa_custom_intent_classification/transformer_classifier.py", line 60, in __init__
    self.model_name = component_config.get("model_name", "albert-base-v2")
AttributeError: 'NoneType' object has no attribute 'get'
```

**Review by nsorros — CHANGES_REQUESTED** [review]
This is great, lets just fix the problem with rasa train and remove the files that Matt pointed

**Comment by pdan93** [comment]
> Are these files meant to be here?
> 
> custom_model/runs/Jan19_14-18-03_daniel-RM-DESKTOP-ECOQUIET-320/1642594687.8485427/events.out.tfevents.1642594687.daniel-RM-DESKTOP-ECOQUIET-320.14899.1 custom_model/runs/Jan19_14-18-03_daniel-RM-DESKTOP-ECOQUIET-320/events.out.tfevents.1642594687.daniel-RM-DESKTOP-ECOQUIET-320.14899.0
> 
> Also I thought this was about sklearn, not about transformers? Is the plan to publish a transformers example too?

solved, and yes, the plan is to  show example for transformers as well.

**Comment by pdan93** [comment]
> `rasa train` on a new env breaks
> 
> ```
> 2022-01-20 09:11:14 INFO     rasa.nlu.model  - Finished training component.
> tokenizer config file saved in /tmp/tmpwlbb5tbm/nlu/tokenizer_component_0_transformer_classifier/tokenizer_config.json
> Special tokens file saved in /tmp/tmpwlbb5tbm/nlu/tokenizer_component_0_transformer_classifier/special_tokens_map.json
> Configuration saved in /tmp/tmpwlbb5tbm/nlu/model_component_0_transformer_classifier/config.json
> Model weights saved in /tmp/tmpwlbb5tbm/nlu/model_component_0_transformer_classifier/pytorch_model.bin
> Configuration saved in /tmp/tmpwlbb5tbm/nlu/config_component_0_transformer_classifier/config.json
> 2022-01-20 09:11:14 INFO     rasa.nlu.model  - Successfully saved model into '/tmp/tmpwlbb5tbm/nlu'
> NLU model training completed.
> Traceback (most recent call last):
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/bin/rasa", line 8, in <module>
>     sys.exit(main())
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/__main__.py", line 117, in main
>     cmdline_arguments.func(cmdline_arguments)
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/cli/train.py", line 59, in <lambda>
>     train_parser.set_defaults(func=lambda args: run_training(args, can_exit=True))
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/cli/train.py", line 91, in run_training
>     training_result = train_all(
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/api.py", line 109, in train
>     return rasa.utils.common.run_in_loop(
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/utils/common.py", line 296, in run_in_loop
>     result = loop.run_until_complete(f)
>   File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 108, in train_async
>     return await _train_async_internal(
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 288, in _train_async_internal
>     await _do_training(
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 358, in _do_training
>     interpreter=_load_interpreter(interpreter_path)
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/model_training.py", line 382, in _load_interpreter
>     return rasa.core.interpreter.create_interpreter(interpreter_path)
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/core/interpreter.py", line 33, in create_interpreter
>     return RasaNLUInterpreter(model_directory=obj)
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/core/interpreter.py", line 127, in __init__
>     self._load_interpreter()
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/core/interpreter.py", line 165, in _load_interpreter
>     self.interpreter = Interpreter.load(self.model_directory)
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/model.py", line 341, in load
>     return Interpreter.create(
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/model.py", line 416, in create
>     component = component_builder.load_component(
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/components.py", line 902, in load_component
>     component = registry.load_component_by_meta(
>   File "/root/code/rasa_custom_intent_classification/build/virtualenv/lib/python3.8/site-packages/rasa/nlu/registry.py", line 177, in load_component_by_meta
>     return component_class.load(
>   File "/root/code/rasa_custom_intent_classification/transformer_classifier.py", line 255, in load
>     x = cls()
>   File "/root/code/rasa_custom_intent_classification/transformer_classifier.py", line 60, in __init__
>     self.model_name = component_config.get("model_name", "albert-base-v2")
> AttributeError: 'NoneType' object has no attribute 'get'
> ```

Hey @nsorros, so I tried to replicate this but all went well when training. Can you try again?

File: requirements.txt (hunk 2 above)
Comment (pdan93) [reply requirements.txt:1]: so, I don't really get what you mean here? (I didn't use our whole unpinned flow, just provided a basic requirements.txt)

Comment (nsorros) [reply requirements.txt:1]: 👍

File: sklearn_classifier.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        requires = ["text"]
+        defaults = {}
+        language_list = ["en"]
+
+        def __init__(self, component_config=None):
+                super().__init__(component_config)
+
+        def _define_model(self, cfg):
+                params = cfg.pipeline[0]
+                kernel, coef0, degree, class_weight = params["kernel"], params["coef0"], params["degree"], params["class_weight"]
```
Comment (nsorros) [inline sklearn_classifier.py]: I assume there are more params than these, otherwise you could simple do `SVC(**params)`

**Comment by nsorros** [comment]
> Hey @nsorros, so I tried to replicate this but all went well when training. Can you try again?

same error

**Comment by nsorros** [comment]
I am also getting these errors when creating the virtualenv
```
ERROR: rasa-sdk 2.8.4 has requirement typing-extensions<4.0.0,>=3.7.4, but you'll have typing-extensions 4.0.1 which is incompatible.
ERROR: rasa-sdk 2.8.4 has requirement uvloop<0.15.0; sys_platform != "win32", but you'll have uvloop 0.16.0 which is incompatible.
ERROR: tensorflow 2.3.4 has requirement gast==0.3.3, but you'll have gast 0.5.3 which is incompatible.
ERROR: sanic-jwt 1.7.0 has requirement pyjwt~=2.1.0, but you'll have pyjwt 2.3.0 which is incompatible.
```

**Comment by pdan93** [comment]
> 

fixed it, try now

File: sklearn_classifier.py (hunk 3 above)
Comment (pdan93) [reply sklearn_classifier.py]: hm, yeah, now that you focus on this I don't like the way it does the `cfg.pipeline[0]` thing, will clean that part differently

**Review by nsorros — APPROVED** [review]

File: sklearn_classifier.py (hunk 3 above)
Comment (pdan93) [reply sklearn_classifier.py]: fixed it so that id doesn't do that weird config thing.

**Pull request merged**